core/src/main/java/org/jdbi/v3/core/internal/QualifiedEnumMapperFactory.java

```diff
@@ -0,0 +1,117 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.internal;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.jdbi.v3.core.Enums;
+import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.generic.GenericTypes;
+import org.jdbi.v3.core.mapper.ColumnMapper;
+import org.jdbi.v3.core.mapper.QualifiedColumnMapperFactory;
+import org.jdbi.v3.core.qualifier.QualifiedType;
+import org.jdbi.v3.core.result.UnableToProduceResultException;
+import org.jdbi.v3.core.statement.StatementContext;
+
+public class QualifiedEnumMapperFactory implements QualifiedColumnMapperFactory {
+    private static final Map<Class<? extends Enum<?>>, ColumnMapper<? extends Enum<?>>> BY_NAME_MAPPER_CACHE = new ConcurrentHashMap<>();
+
+    @Override
+    public Optional<ColumnMapper<?>> build(QualifiedType<?> givenType, ConfigRegistry config) {
+        return Optional.of(givenType.getType())
+            .map(GenericTypes::getErasedType)
+            .filter(Class::isEnum)
+            .flatMap(clazz -> makeEnumArgument((QualifiedType<Enum>) givenType, (Class<Enum>) clazz, config));
+    }
+
+    private static <E extends Enum<E>> Optional<ColumnMapper<?>> makeEnumArgument(QualifiedType<E> givenType, Class<E> enumClass, ConfigRegistry config) {
+        boolean byName = Enums.EnumStrategy.BY_NAME == config.get(Enums.class).findStrategy(givenType, enumClass);
+
+        return Optional.of(byName
+            ? byName(enumClass)
+            : byOrdinal(enumClass));
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <E extends Enum<E>> ColumnMapper<E> byName(Class<E> enumClass) {
+        return (ColumnMapper<E>) BY_NAME_MAPPER_CACHE.computeIfAbsent(enumClass, e -> new EnumByNameColumnMapper<>(enumClass));
+    }
+
+    public static <E extends Enum<E>> ColumnMapper<E> byOrdinal(Class<E> enumClass) {
+        return new EnumByOrdinalColumnMapper<>(enumClass);
+    }
+
+    private static class EnumByNameColumnMapper<E extends Enum<E>> implements ColumnMapper<E> {
+        private final Map<String, E> nameValueCache = new ConcurrentHashMap<>();
+        private final Class<E> enumClass;
+
+        private EnumByNameColumnMapper(Class<E> enumClass) {
+            this.enumClass = enumClass;
+        }
+
+        @Override
+        public E map(ResultSet rs, int columnNumber, StatementContext ctx) throws SQLException {
+            String name = ctx.findColumnMapperFor(String.class)
+                .orElseThrow(() -> new UnableToProduceResultException("a String column mapper is required to map Enums from names", ctx))
+                .map(rs, columnNumber, ctx);
```

Comment:
Same as the other benchmark, except now I'm comparing mapper lookup vs `getString()` / `getInt()`.

Using `getString()` / `getInt()`:

```
Benchmark                                Mode  Cnt      Score     Error  Units
EnumMapperBenchmark.mapByExactName       avgt    5  79775.267 ± 460.790  ns/op
EnumMapperBenchmark.mapByOrdinal         avgt    5  74462.557 ± 546.199  ns/op
EnumMapperBenchmark.mapByRandomCaseName  avgt    5  92999.559 ± 447.022  ns/op
```

Using column mapper lookup from registry:

```
EnumMapperBenchmark.mapByExactName       avgt    5  123372.656 ±  823.765  ns/op
EnumMapperBenchmark.mapByOrdinal         avgt    5  118028.199 ±  450.061  ns/op
EnumMapperBenchmark.mapByRandomCaseName  avgt    5  137851.096 ± 2441.665  ns/op
```

We're paying a hefty performance penalty for the mapper lookup--55%, 59%, and 48%, respectively.

We could EnumByName mapper to check for empty string:

```
Benchmark                                Mode  Cnt      Score      Error  Units
EnumMapperBenchmark.mapByExactName       avgt    5  81420.110 ± 3290.003  ns/op
EnumMapperBenchmark.mapByOrdinal         avgt    5  75045.948 ±  298.406  ns/op
EnumMapperBenchmark.mapByRandomCaseName  avgt    5  91684.408 ±  416.760  ns/op
```

This is roughly on par with straight `getString()`. Full disclosure: I had tweak the benchmark for this last run, so that it would insert nulls randomly into the database.

@TheRealMarnes Would you be OK dropping the mapper lookup, if we check for empty string and treat it as null for enum mapping purposes?

Reply:
I think the factory should throw an exception on empty strings because they can't ever match any enum value we could find